src/app/clusters/zone-management-server/zone-management-server.cpp

```diff
@@ -0,0 +1,826 @@
+    {
+        ChipLogError(Zcl, "HandleUpdateTwoDCartesianZone: Found self-intersecting polygon vertices for zone");
+        ctx.mCommandHandler.AddStatus(ctx.mRequestPath, Status::DynamicConstraintError);
+        return;
+    }
+
+    TwoDCartesianZoneStorage twoDCartZoneStorage;
+    twoDCartZoneStorage.Set(zoneToUpdate.name, zoneToUpdate.use, twoDCartVertices, zoneToUpdate.color);
+
+    // Call the delegate
+    status = mDelegate.UpdateTwoDCartesianZone(zoneID, twoDCartZoneStorage);
+    if (status != Status::Success)
+    {
+        ctx.mCommandHandler.AddStatus(ctx.mRequestPath, status);
+        return;
+    }
+
+    ZoneInformationStorage zoneInfo;
+    zoneInfo.Set(zoneID, ZoneTypeEnum::kTwoDCARTZone, ZoneSourceEnum::kUser, MakeOptional(twoDCartZoneStorage));
+
+    UpdateZone(zoneID, zoneInfo);
+
+    ctx.mCommandHandler.AddStatus(ctx.mRequestPath, Status::Success);
+}
+
+void ZoneMgmtServer::HandleRemoveZone(HandlerContext & ctx, const Commands::RemoveZone::DecodableType & commandData)
+{
+    uint16_t zoneID = commandData.zoneID;
+
+    auto foundZone =
+        std::find_if(mZones.begin(), mZones.end(), [&](const ZoneInformationStruct & z) { return z.zoneID == zoneID; });
+    if (foundZone == mZones.end())
+    {
+        ChipLogError(Zcl, "ZoneMgmt[ep=%d]: No zone exists by id %d", mEndpointId, zoneID);
+        ctx.mCommandHandler.AddStatus(ctx.mRequestPath, Status::NotFound);
+        return;
+    }
+
+    if (foundZone->zoneSource == ZoneSourceEnum::kMfg)
+    {
+        ctx.mCommandHandler.AddStatus(ctx.mRequestPath, Status::ConstraintError);
+        return;
+    }
+
+    auto foundTrigger =
+        std::find_if(mTriggers.begin(), mTriggers.end(), [&](const ZoneTriggerControlStruct & t) { return t.zoneID == zoneID; });
+    if (foundTrigger != mTriggers.end())
+    {
+        ChipLogError(Zcl, "ZoneMgmt[ep=%d]: Zone id %d is referenced in zone triggers", mEndpointId, zoneID);
+        ctx.mCommandHandler.AddStatus(ctx.mRequestPath, Status::InvalidInState);
+        return;
+    }
+
+    // Call the delegate
+    Status status = mDelegate.RemoveZone(zoneID);
+
+    if (status == Status::Success)
+    {
+        RemoveZone(zoneID);
+        mUserDefinedZonesCount--;
+    }
+
+    ctx.mCommandHandler.AddStatus(ctx.mRequestPath, status);
+}
+
+Status ZoneMgmtServer::ValidateTrigger(const ZoneTriggerControlStruct & trigger)
+{
+    VerifyOrReturnError(trigger.initialDuration >= 1 && trigger.initialDuration <= 65535, Status::ConstraintError);
+
+    VerifyOrReturnError(trigger.augmentationDuration <= trigger.initialDuration, Status::ConstraintError);
+
+    VerifyOrReturnError(trigger.maxDuration >= trigger.initialDuration, Status::ConstraintError);
+
+    if (HasFeature(Feature::kPerZoneSensitivity))
+    {
+        VerifyOrReturnError(trigger.sensitivity.Value() >= 1 && trigger.sensitivity.Value() <= mSensitivityMax,
+                            Status::ConstraintError);
+    }
```

Comment:
This code does not match the spec, and will crash on one variety of invalid input (i.e. when the sensitivity is not present but the feature bit is set).

Reply:
Yes, missed adding the absence/presence validation of that field in the command args. Added that.